best-practices.mdx

```diff
@@ -0,0 +1,190 @@
+from datetime import datetime
+
+@features
+class User:
+  id: str
+
+  # the user's fullname
+  # :owner: mary.shelley@aol.com
+  # :tags: team:identity, priority:high
+  name: str
+
+  # the user's birthdate
+  # :owner: mary.shelley@aol.com
+  # :tags: team:identity, priority:high
+  birthday: datetime
+```
+
+If you want to apply a tag or owner to all features in a feature set, this should be done in the featuer decorator, like so:
+
+```python
+from chalk import features
+from datetime import datetime
+
+@features(owner="ada.lovelace@aol.com", tags=['group:risk'])
+class User:
+  id: str
+
+  # the user's fullname.
+  name: str
+```
+
+You can also [apply restrictions or enforce feature annotation for your entire project](/validation#feature-validation-requirements). For instance, you can block deployment if features are not tagged or described.
+
+### Keep feature definitions separated from resolvers
+
+Features should be defined in separate files from resolvers (with the exception of underscore features).
+
+### Define All Your Features in the Same File
+
+Although it can get a little lengthy, we recommend defining your features in a single file. This makes expressing joins between features easier and prevents circular
+dependencies.
+
+### Add Validations For Your Features
+
+[Validations for your features](/validation#feature-values) can prevent incorrect data from being written to your offline store. They can provide an even stricter complement to monitoring, ensuring that nothing
+is going wrong with the feature you are calculating.
+
+### Use implicit join syntax
+
+Joins between feature sets can be specified in a number of different ways. We recommend using an implicit join syntax, which [we cover in the join section of the docs](/has-one#recommended-implementation).
+
+## Resolvers
+
+### Use underscore resolvers for very simple resolver definitions
+
+[Underscore resolvers](/underscore) should be used for relatively simple calculations: your underscore resolver definitions should fit inline.
+
+### Use SQL file resolvers to read data from your raw datasets
+
+While python resolvers can be used to read data from your data sources, SQL File resolvers are preferred.
+
+### Explicity list columns in a select statement for sql file resolvers
+
+Select statements in SQL file resolvers should be explicit: avoid using the \* syntax.
+
+### Make sure your resolvers operate inside a single feature space
+
+Resolver inputs and outputs must belong to the same feature set, but joins can allow resolvers to connect data between feature sets.
+
+### When to use pandas vs polars transformation.
+
+Don't worry about converting Chalk DataFrames to pandas or polars in a python resolver—the transformation is cheap. We use arrow (and so do pandas and polars ) so moving data from a chalk dataframe to either is close to free
+
+## Querying
+
+### Run Simple Queries With The Chalk CLI, For More Flexibility Use One of Chalk's Clients
```

Comment:
We should probably pick one of Capitalize Every Letter In a Header and Only capitalize the first letter